### PR TITLE
workload/schemachange: classify schema change errors

### LINKED ISSUE #23286: roachtest: table creation stress test
We should put the investigation in https://github.com/cockroachdb/cockroach/issues/23254 into a roachtest. Here's the relevant bit of code courtesy of @andreimatei  (no need to transcribe this verbatim at all, should be enough to follow it in spirit):

```go
func TestParallelDropCreateTables(t *testing.T) {
	defer leaktest.AfterTest(t)()

	// This number has to be around 10 or else testrace will take too long to
	// finish.
	const numberOfNodes = 3
	const workers = 60
	const repeat = 1000

	tc := testcluster.StartTestCluster(t, numberOfNodes, base.TestClusterArgs{})
	defer tc.Stopper().Stop(context.TODO())

	if _, err := tc.ServerConn(0).Exec(`CREATE DATABASE "test"`); err != nil {
		t.Fatal(err)
	}

	var wgStart sync.WaitGroup
	var wgEnd sync.WaitGroup
	wgStart.Add(workers)
	wgEnd.Add(workers)
	signal := make(chan struct{})
	for i := 0; i < workers; i++ {
		db := tc.ServerConn(i % numberOfNodes)
		go createTestTable2(t, tc, i, repeat, db, &wgStart, &wgEnd, signal)
	}

	// Wait until all goroutines are ready.
	wgStart.Wait()
	// Signal the create table goroutines to start.
	close(signal)
	// Wait until all create tables are finished.
	wgEnd.Wait()

}
func createTestTable2(
	t *testing.T,
	tc *testcluster.TestCluster,
	id int,
	repeat int,
	db *gosql.DB,
	wgStart *sync.WaitGroup,
	wgEnd *sync.WaitGroup,
	signal chan struct{},
) {
	defer wgEnd.Done()

	wgStart.Done()
	<-signal

	for i := 0; i < repeat; i++ {
		log.Infof(context.TODO(), "!!! worker %d loop %d", id, i)
		tableName := fmt.Sprintf("%d_%d", id, i)
		tableSQL := fmt.Sprintf(`
		CREATE TABLE "test"."table_%s" (
			id INT PRIMARY KEY,
			val INT
		)`, tableName)
		// if i%2 == 0 {
		//   txn, err := db.Begin()
		//   if err != nil {
		//     t.Errorf("table %d: could not be created. begin err: %s", id, err)
		//     panic("!!!")
		//     return
		//   }
		//   if _, err := txn.Exec(tableSQL); err != nil {
		//     t.Errorf("table %d: could not be created: %s", id, err)
		//     panic("!!!")
		//     return
		//   }
		//   if err := txn.Commit(); err != nil {
		//     t.Errorf("table %d: could not be created. commit err: %s", id, err)
		//     panic("!!!")
		//     return
		//   }
		// } else {
		if _, err := db.Exec(tableSQL); err != nil {
			t.Errorf("table %d: could not be created: %s", id, err)
			return
		}
		// }
	}
}

```

### PR DESCRIPTION
**Pull request by spaskob**
Previously this workload will not fail even when a schema change
operation returns an error. Now we extract the SQLSTATE code from
the error and if it is of class `09`, i.e. Triggered Action Exception
or `XX`, i.e. Internal Error we abort the workload and report the error.
Also previously we aborted the txn at the first operation error, now
we continue unless the error is one of the classes above.
We also add a roachtest `schemachange/random-load` that runs
the workload and fails if unclassified errors are encountered.

Fixes #23286.

Release note: none.

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6UnjwC03UO5cJYTdos:-M6UnjwC03UO5cJYTdot:b303wfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1be2908641f386cb65c4c8fe06910745e6a514da/pkg/workload/schemachange/schemachange.go#L274)):*
> ```Go
> }
> 
> func (w *schemaChangeWorker) runInTxn(tx *pgx.Tx) (string, bool, error) {
> ```

I have mixed feelings about all of the string inspection happening here. Have you trying pulling a [pgconn.PgError](https://godoc.org/github.com/jackc/pgconn#PgError) out of the error returned from `tx.Exec` and then pulling its code out?

**Review by spaskob — COMMENTED** [review]
Addressed the feedback and fixed the TC failures are due the fact that the workload actually finds current issues and fails so this PR will have to be merged after I fix the long tail of these problems. They seem to be related to my recent changed of validation - certain validations for duplicates don't get triggered and we hit the assertion failures in ValidateTable but at least it;'s working as expected and is finding bugs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-:-M6_g0NyBTL6TjWcS3W0:byi9a28)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/cmd/roachtest/schemachange_random_load.go, line 48 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oPnSmAKadRxKCmEu4:-M6oPnSnBWAICgrbFLSO:bbln1q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/cmd/roachtest/schemachange_random_load.go#L48)):*
> ```Go
> 	runCmd := []string{
> 		"./workload run",
> 		fmt.Sprintf("schemachange  --concurrency %d --max-ops %d --verbose=1", maxOps,
> ```

These arguments seem backwards

Also seems weird that this is just a single string. 
___
*[pkg/workload/schemachange/schemachange.go, line 247 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oR7ubDkmonXHxwdSs:-M6oR7ubDkmonXHxwdSt:b-whzhst) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6186c744a92ed7329d6be213f3a7c4d098c8df0/pkg/workload/schemachange/schemachange.go#L247)):*
> ```Go
> }
> 
> func (w *schemaChangeWorker) runInTxn(tx *pgx.Tx) (string, error) {
> ```

This rollback and error classification seems like it might be happening at the wrong level. It seems like any error needs to be dealt with for the whole transaction. It feels like this function should just return the error and then run should rollback on probably all of the errors but then decide whether to tolerate the error or not. 
___
*[pkg/workload/schemachange/schemachange.go, line 272 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oQE1U4Azt5Y7Wh1l-:-M6oQE1V8U5lSrN4OIDU:b-b6u892) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/workload/schemachange/schemachange.go#L272)):*
> ```Go
> }
> 
> func (w *schemaChangeWorker) runInTxn(tx *pgx.Tx) (string, bool, error) {
> ```

Can you name the bool return param?
___
*[pkg/workload/schemachange/schemachange.go, line 304 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oR2gw9miJTDXGAXMg:-M6oR2gw9miJTDXGAXMh:b-mj6r7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/workload/schemachange/schemachange.go#L304)):*
> ```Go
> 		if seriousErr := handleOpError(err); seriousErr != nil {
> 			if w.tolerateErrors {
> 				opsBuf.WriteString(fmt.Sprintf("Serious error: %v\n", seriousErr))
> ```

Don’t you still want to bail out of this transaction?
___
*[pkg/workload/schemachange/schemachange.go, line 308 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oQd_D6btNhAxrXwh8:-M6oQd_D6btNhAxrXwh9:b-4pzeg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/workload/schemachange/schemachange.go#L308)):*
> ```Go
> 				return opsBuf.String(), rollback, seriousErr
> 			}
> 		}
> ```

If rollback ever becomes true, do you really want to proceed to the next iteration of the loop? The next command is likely to fail now that you’re in an error state.

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/cmd/roachtest/schemachange_random_load.go, line 48 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oPnSmAKadRxKCmEu4:-M6otoQZ5YCkp5UqXB36:b-ena300) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/cmd/roachtest/schemachange_random_load.go#L48)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

These arguments seem backwards

Also seems weird that this is just a single string. 
</blockquote></details>

good catch - made it more maintainable
___
*[pkg/workload/schemachange/schemachange.go, line 247 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oR7ubDkmonXHxwdSs:-M6ouQ6FE4KsjNbP3Vbo:b-51t7p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6186c744a92ed7329d6be213f3a7c4d098c8df0/pkg/workload/schemachange/schemachange.go#L247)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This rollback and error classification seems like it might be happening at the wrong level. It seems like any error needs to be dealt with for the whole transaction. It feels like this function should just return the error and then run should rollback on probably all of the errors but then decide whether to tolerate the error or not. 
</blockquote></details>

well I tried alternatives before coming up with this and also considered what you are suggesting - the trouble is we are still not clear on what errors we want to catch exactly and until we have some clarity as a team on what ultimately should be the purpose of this tool, I prefer to leave it as it is and not sink too much cost upfront. A few open questions:
- do we want to abort txn on first failure or continue to understand failure scenarios under cascaded errors?
- what kind of errors do we want to generate on purpose
- how to actually find out if something broke - currently we don't and we need some kind of database internals consistency checks after each txn commits or is rolled back

So there is too much uncertainty about this tool so I'd rather punt on the issue your highlighting for now.
___
*[pkg/workload/schemachange/schemachange.go, line 272 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oQE1U4Azt5Y7Wh1l-:-M6otxwW6Tx4l6pYuwDW:b5iem5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/workload/schemachange/schemachange.go#L272)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you name the bool return param?
</blockquote></details>

What do you mean? Use a named return value? I thought that was discouraged unless it's really necessary or you are asking for a comment for the function?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 247 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oR7ubDkmonXHxwdSs:-M6pOe5IDzi6nRCGXgwG:b-oz9ss3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6186c744a92ed7329d6be213f3a7c4d098c8df0/pkg/workload/schemachange/schemachange.go#L247)):*
> do we want to abort txn on first failure or continue to understand failure scenarios under cascaded errors?

In SQL once you get an error, unless you `ROLLBACK TO SAVEPOINT` your transaction is toast and all future statements are invalid. Given this fact, I'm not sure the rest of your points make as much sense. If you want the higher level to be able to cope with the error better, we should wrap the surrounding context and pass it up the stack. Dealing with the error here makes the code hard to follow. It seems to me that this lower level should inform the high level of the relevant context of the error but should leave things like rollback decisions to the higher level based on the context communicated from the lower level.

Maybe in the future we introduce savepoints in the `run` call, I think that's probably an interesting space for bugs. 
___
*[pkg/workload/schemachange/schemachange.go, line 272 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oQE1U4Azt5Y7Wh1l-:-M6pOmIVCTCZtblcoQUl:b-fyke2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/772156c5758747d2c3620dd7388d30c171cce386/pkg/workload/schemachange/schemachange.go#L272)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

What do you mean? Use a named return value? I thought that was discouraged unless it's really necessary or you are asking for a comment for the function?
</blockquote></details>

It's encouraged to use the named return values, it makes reading signatures easier. It's highly discouraged to do naked returns (which are enabled by using return values).

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 247 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oR7ubDkmonXHxwdSs:-M7c38dr8-EXVMSvstDL:bwfykf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6186c744a92ed7329d6be213f3a7c4d098c8df0/pkg/workload/schemachange/schemachange.go#L247)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> do we want to abort txn on first failure or continue to understand failure scenarios under cascaded errors?

In SQL once you get an error, unless you `ROLLBACK TO SAVEPOINT` your transaction is toast and all future statements are invalid. Given this fact, I'm not sure the rest of your points make as much sense. If you want the higher level to be able to cope with the error better, we should wrap the surrounding context and pass it up the stack. Dealing with the error here makes the code hard to follow. It seems to me that this lower level should inform the high level of the relevant context of the error but should leave things like rollback decisions to the higher level based on the context communicated from the lower level.

Maybe in the future we introduce savepoints in the `run` call, I think that's probably an interesting space for bugs. 


</blockquote></details>

I agree about terminating the txn on first error - you are right. As for handling the errors in the caller, here is my point this function has 2 type of errors:
- ones that force terminating the whole workload:
    - unable to create a new op
    - having encountered a serious error (this one is debatable as we may still want to proceed, save the serious error and report to the user in the end)
- ones that just force rollback but the workload continues.

The caller will needs to distinguish b/w the two types and also create the log of operations which kinda depends on the guts of this function. So I am leaning towards rolling back in side this function if we have to and the caller just dealing with questions around should we terminate the whole workload.

WDYT?

**Comment by ajwerner** [comment]
> So I am leaning towards rolling back in side this function if we have to and the caller just dealing with questions around should we terminate the whole workload.

That sounds good to me.

**Comment by spaskob** [comment]
PTAL

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r6, 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-:-M8L_3BmDym-BN1Pq_uX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 247 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M6oR7ubDkmonXHxwdSs:-M8LYBQU-In8GfMeeMiI:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6186c744a92ed7329d6be213f3a7c4d098c8df0/pkg/workload/schemachange/schemachange.go#L247)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

I agree about terminating the txn on first error - you are right. As for handling the errors in the caller, here is my point this function has 2 type of errors:
- ones that force terminating the whole workload:
    - unable to create a new op
    - having encountered a serious error (this one is debatable as we may still want to proceed, save the serious error and report to the user in the end)
- ones that just force rollback but the workload continues.

The caller will needs to distinguish b/w the two types and also create the log of operations which kinda depends on the guts of this function. So I am leaning towards rolling back in side this function if we have to and the caller just dealing with questions around should we terminate the whole workload.

WDYT? 
</blockquote></details>

:+1: 
___
*[pkg/workload/schemachange/schemachange.go, line 293 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M8LX54g1pp6Glrwci0K:-M8LX54g1pp6Glrwci0L:btyzmsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/334526ed9526b0b8d9713680a256f277011b2d6a/pkg/workload/schemachange/schemachange.go#L293)):*
> ```Go
> 			if _, err = tx.Exec(op); err != nil {
> 				histBin = "txnRbk"
> 				log.WriteString(fmt.Sprintf("***FAIL: %v\n", err))
> ```

Should this `***FAIL` be so loud here? Don't we expect these pretty frequently? It is good to log the error, just seems to indicate that 


Also, maybe it makes sense to append the `ROLLBACK` to the logs in `run` and similarly add `COMMIT` upon success there as well.

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48338)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/workload/schemachange/schemachange.go, line 293 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/48338#-M8LX54g1pp6Glrwci0K:-M8Lf6z62J78yksBaMb3:bd3re65) ([raw file](https://github.com/cockroachdb/cockroach/blob/334526ed9526b0b8d9713680a256f277011b2d6a/pkg/workload/schemachange/schemachange.go#L293)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should this `***FAIL` be so loud here? Don't we expect these pretty frequently? It is good to log the error, just seems to indicate that 


Also, maybe it makes sense to append the `ROLLBACK` to the logs in `run` and similarly add `COMMIT` upon success there as well. 
</blockquote></details>

This is at verbosity level 1 and above so only loud if you ask for it - I find it helpful for grepping and classifying the errors. At verbosity 0 only serious errors will be printed.
As for the sites where to ROLLBACK; I've tried both more than once, there's always pros and cons. I'd rather not touch it again.
